Proposals/NNNN-concurrency-safe-notifications.md

```diff
@@ -0,0 +1,386 @@
+}
+```
+
+`NotificationCenter.Message` is designed to interoperate with existing uses of `Notification` by sharing `Notification.Name` identifiers. This means an observer expecting `NotificationCenter.Message` will be called when a `Notification` is posted if the `Notification.Name` identifier matches, and vice versa.
+
+The protocol specifies `makeMessage(:Notification)` and `makeNotification(:Self)` to transform the payload between posters and observers of both the `NotificationCenter.Message` and `Notification` types. These methods have default implementations in cases where interoperability with `Notification` is not necessary.
+
+For `Message` types that do not need to interoperate with existing `Notification` uses, the `name` property does not need to be specified, and will default to the fully qualified name of the `Message` type, e.g. `MyModule.MyMessage`. Note that when using this default, renaming the type or relocating it to another module has a similar effect as changing ABI, as any code that was compiled separately will not be aware of the name change until recompiled. Developers can control this effect by explicitly setting the `name` property if needed.
+
+### Observing messages
+
+Observing messages can be done with new overloads to `addObserver`. Clients do not need to know whether a message conforms to `MainActorMessage` or `AsyncMessage`.
+
+For `MainActorMessage`:
+
+```swift
+@available(FoundationPreview 0.5, *)
+extension NotificationCenter {
+    // e.g. addObserver(of: workspace, for: .willLaunchApplication) { message in ... }
+    public func addObserver<I: MessageIdentifier, M: MainActorMessage>(of subject: M.Subject,
+                                                                       for identifier: I,
+                                                                       using observer: @escaping @MainActor (M) -> Void)
+        -> ObservationToken where I.MessageType == M
+
+    // e.g. addObserver(of: NSWorkspace.self, for: .willLaunchApplication) { message in ... }
+    public func addObserver<I: MessageIdentifier, M: MainActorMessage>(of subject: M.Subject.Type,
+                                                                       for identifier: I,
+                                                                       using observer: @escaping @MainActor (M) -> Void)
+        -> ObservationToken where I.MessageType == M
+
+    // e.g. addObserver(NSWorkspace.WillLaunchApplication.self) { message in ... }
+    public func addObserver<M: MainActorMessage>(_ messageType: M.Type,
+                                                 subject: M.Subject? = nil,
+                                                 using observer: @escaping @MainActor (M) -> Void)
+        -> ObservationToken
+}
+```
+
+And for `AsyncMessage`:
+
+```swift
+@available(FoundationPreview 0.5, *)
+extension NotificationCenter {
+    public func addObserver<I: MessageIdentifier, M: AsyncMessage>(of subject: M.Subject,
```

Comment:
How can those methods take an `async` closure while not being async themselves? Where is the task spawned that calls the observer closures?

If the `addObserver` method takes an async closure I would have expected that method to become `async` itself. Something more like this

```
await center.observe(of: subject, for: identifier) { message
  // do something async with the message
  await process(message)
}
```

This way you also don't need `ObservationToken`s but can just rely on task cancellation which IMO fits better into the Concurrency paradigms.